### PR TITLE
Add tensor conversion to `flatnonzero`, `nonzero_values`, `tile`, `inverse_permutation`, and `diag`

### PR DESCRIPTION
**Pull request by ltoniazzi**
Fixing [Issue](https://github.com/aesara-devs/aesara/issues/876) by adding tensor conversion to Numpy-like helper functions. Tests added by passing a list/float to these functions.

Functions addressed:
- `flatnonzero`
- `diag`
- `inverse_permutation`
- `tile`
- `identity_like`

**Thank you for opening a PR!**

Here are a few important guidelines and requirements to check before your PR can be merged:
+ [x] There is an informative high-level description of the changes.
+ [x] The description and/or commit message(s) references the relevant GitHub issue(s).
+ [x] [`pre-commit`](https://pre-commit.com/#installation) is installed and [set up](https://pre-commit.com/#3-install-the-git-hook-scripts).
+ [ ] The commit messages follow [these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [ ] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes), and there are **no commits that fix changes introduced by other commits in the same branch/BR**.
+ [x] There are tests covering the changes introduced in the PR.

Don't worry, your PR doesn't need to be in perfect order to submit it.  As development progresses and/or reviewers request changes, you can always [rewrite the history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_rewriting_history) of your feature/PR branches.

If your PR is an ongoing effort and you would like to involve us in the process, simply make it a [draft PR](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests).

### CODE REVIEW

File: aesara/tensor/basic.py [hunk 1]
```diff
@@ -3575,12 +3579,14 @@ def diag(v, k=0):
...
-        return diagonal(v, offset=k)
+    _v = as_tensor_variable(v)
+
+    if _v.ndim == 1:
+        return AllocDiag(k)(_v)
+    elif _v.ndim >= 2:
+        return diagonal(_v, offset=k)
     else:
-        raise ValueError("Input must has v.ndim >= 1.")
+        raise ValueError("Input must have _v.ndim >= 1.")
```
Comment (brandonwillard) [inline aesara/tensor/basic.py:3589]:
```suggestion
        raise ValueError("Number of dimensions of `v` must be greater than one.")
```

File: tests/tensor/test_basic.py [hunk 2]
```diff
@@ -1023,6 +1023,12 @@ def check(m):
         rand2d[:4] = 0
         check(rand2d)
 
+        # Test passing a list
+        m = [1, 2, 0]
+        out = flatnonzero(m)
+        f = function([], out)
+        assert np.all(f() == np.flatnonzero(m))
```
Comment (brandonwillard) [inline tests/tensor/test_basic.py:1030]:
```suggestion
        assert np.array_equal(f(), np.flatnonzero(m))
```

File: tests/tensor/test_basic.py [hunk 3]
```diff
@@ -2221,6 +2226,16 @@ def run_tile(x, x_, reps, use_symbolic_reps):
...
 
+        # Test passing a float
+        x = scalar()
+        x_ = 1.0
+        assert np.all(run_tile(x, x_, (2,), use_symbolic_reps) == np.tile(x_, (2,)))
+
+        # Test when x is a list
+        x = matrix()
+        x_ = [[1.0, 2.0], [3.0, 4.0]]
+        assert np.all(run_tile(x, x_, (2,), use_symbolic_reps) == np.tile(x_, (2,)))
```
Comment (brandonwillard) [inline tests/tensor/test_basic.py:2237]:
```suggestion
        x_val = [[1.0, 2.0], [3.0, 4.0]]
        assert np.array_equal(run_tile(x, x_val, (2,), use_symbolic_reps), np.tile(x_val, (2,)))
```

File: tests/tensor/test_basic.py [hunk 4]
```diff
@@ -2221,6 +2226,16 @@ def run_tile(x, x_, reps, use_symbolic_reps):
             == np.tile(x_, (2, 3, 4, 6))
         )
 
+        # Test passing a float
+        x = scalar()
+        x_ = 1.0
+        assert np.all(run_tile(x, x_, (2,), use_symbolic_reps) == np.tile(x_, (2,)))
```
Comment (brandonwillard) [inline tests/tensor/test_basic.py:2232]:
```suggestion
        x_val = 1.0
        assert np.array_equal(run_tile(x, x_val, (2,), use_symbolic_reps), np.tile(x_val, (2,)))
```

File: tests/tensor/test_basic.py [hunk 5]
```diff
@@ -1449,8 +1455,7 @@ def test_roll(self):
 
             assert (out == want).all()
 
-            # Pass a list to make sure `a` is converted to a
-            # TensorVariable by roll
+            # Test rolling passing a list
```
Comment (brandonwillard) [inline tests/tensor/test_basic.py:1458]:
```suggestion
```

File: tests/tensor/test_basic.py [hunk 6]
```diff
@@ -2794,6 +2809,12 @@ def test_dim1(self):
         assert np.all(p_val[inv_val] == np.arange(10))
         assert np.all(inv_val[p_val] == np.arange(10))
 
+        # Test passing a list
+        p = [2, 4, 3, 0, 1]
+        inv = at.inverse_permutation(p)
+        f = aesara.function([], inv)
+        assert np.all(f() == np.array([3, 4, 0, 2, 1]))
```
Comment (brandonwillard) [inline tests/tensor/test_basic.py:2816]:
```suggestion
        assert np.array_equal(f(), np.array([3, 4, 0, 2, 1]))
```

File: tests/tensor/test_basic.py [hunk 7]
```diff
@@ -3449,6 +3470,12 @@ def test_diag(self):
         with pytest.raises(ValueError):
             diag(xx)
 
+        # Test passing a list
+        xx = [[1, 2], [3, 4]]
+        g = diag(xx)
+        f = function([], g)
+        assert np.all(f() == np.diag(xx))
```
Comment (brandonwillard) [inline tests/tensor/test_basic.py:3477]:
```suggestion
        assert np.array_equal(f(), np.diag(xx))
```

File: tests/tensor/test_basic.py [hunk 8]
```diff
@@ -4136,6 +4163,12 @@ def test_identity_like_dtype():
     m_out_float = identity_like(m, dtype=np.float64)
     assert m_out_float.dtype == "float64"
 
+    # Test passing list
+    m = [[0, 1], [1, 3]]
+    out = at.identity_like(m)
+    f = aesara.function([], out)
+    assert np.all(f() == np.eye(2))
```
Comment (brandonwillard) [inline tests/tensor/test_basic.py:4170]:
```suggestion
    assert np.array_equal(f(), np.eye(2))
```

**Review by brandonwillard — CHANGES_REQUESTED** [review]

**Comment by ltoniazzi** [comment]
Sorry, I merged instead of rebasing, is that ok?

**Comment by brandonwillard** [comment]
> Sorry, I merged instead of rebasing, is that ok?

You'll need to rebase instead.

**Comment by ltoniazzi** [comment]
Should I squash the commits now?

**Pull request merged**